packages/tpetra/core/test/Utils/TpetraUtils_crsUtils.cpp

```diff
@@ -0,0 +1,309 @@
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+*/
+
+#include "Tpetra_TestingUtilities.hpp"
+#include "Tpetra_Core.hpp"
+#include "Tpetra_Map.hpp"
+#include "Tpetra_ConfigDefs.hpp"
+#include "Tpetra_Details_crsUtils.hpp"
+
+#include "Kokkos_Core.hpp"
+#include "Teuchos_UnitTestHarness.hpp"
+#include "Teuchos_Array.hpp"
+#include "Teuchos_as.hpp"
+
+#include <algorithm>
+#include <iterator>
+
+using Teuchos::CommandLineProcessor;
+using Tpetra::Details::padCrsArrays;
+using Tpetra::Details::insertCrsIndices;
+using Tpetra::Details::impl::ind_difference;
+using std::vector;
+
+namespace {
+
+  TEUCHOS_STATIC_SETUP()
+  {
+    Teuchos::CommandLineProcessor &clp = Teuchos::UnitTestRepository::getCLP();
+    clp.addOutputSetupOptions(true);
+  }
+
+  //
+  // UNIT TESTS
+  //
+TEUCHOS_UNIT_TEST(CrsGraph, ResizeRowPointersAndIndices_1)
+{
+  using device_type = typename Tpetra::Map<>::device_type;
+  using execution_space = typename device_type::execution_space;
+  using ordinal_type = size_t;
+  using view_type = Kokkos::View<ordinal_type*, device_type>;
+  using size_type = typename view_type::size_type;
+
+  ordinal_type num_row = 4;
+  ordinal_type num_indices_per_row = 5;
+  ordinal_type num_indices = num_indices_per_row * num_row;
+  auto row_ptrs_beg = view_type("beg", num_row+1);
+  // this assumes UVM
```

Comment:
You should insert a fence here.  View's constructor will fill on device, asynchronously, so your subsequent writes on host may conflict.

Reply:
Also, consider allocating the View without initializing, since you're going to fill it anyway (and it's a device View).